File: fetch_images.py
import urllib.request
import urllib.parse
import json
import time

def fetch_exact_image(file_name, out_path):
    headers = {'User-Agent': 'MeoncrossHistoryApp/1.0 (test4@test.com)'}
    
    try:
        # Now get the url
        img_req_url = f"https://commons.wikimedia.org/w/api.php?action=query&prop=imageinfo&iiprop=url&titles=File:{urllib.parse.quote(file_name)}&format=json"
        req2 = urllib.request.Request(img_req_url, headers=headers)
        
        response2 = urllib.request.urlopen(req2)
        data2 = json.loads(response2.read().decode('utf-8'))
        
        pages = data2.get('query', {}).get('pages', {})
        page_id = list(pages.keys())[0]
        
        if page_id == "-1":
            print(f"No image info for {file_name}")
            return
            
        img_url = pages[page_id]['imageinfo'][0]['url']
        print(f"Downloading {img_url} to {out_path}...")
        
        req3 = urllib.request.Request(img_url, headers=headers)
        img_data = urllib.request.urlopen(req3).read()
        
        with open(out_path, 'wb') as f:
            f.write(img_data)
            
        print("Success!\n")
        time.sleep(1) # prevent rate limit
        
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    fetch_exact_image('Alsace Lorraine departments evolution map-en.svg', 'great_war/assets/alsace_lorraine_simple_map.svg')
    fetch_exact_image('Sir J. Harington, Diagram of the water-closet Wellcome M0012969.jpg', 'water_and_sanitation/assets/harington_toilet.jpg')
    fetch_exact_image('Snow-cholera-map-1.jpg', 'water_and_sanitation/assets/snow_cholera_map.jpg')
    fetch_exact_image('London Sewage system being built in 1860.jpg', 'water_and_sanitation/assets/bazalgette_sewer.jpg')

File: fetch_images.js
const https = require('https');
const fs = require('fs');
const path = require('path');

const imagesToFetch = [
  { title: "Golda_Meir", filename: "golda_meir.jpg" },
  { title: "Yitzhak_Shamir", filename: "yitzhak_shamir.jpg" },
  { title: "Hussein_of_Jordan", filename: "king_hussein.jpg" },
  { title: "Oslo_Accords", filename: "oslo_handshake.jpg" }
];

async function fetchThumbnailUrl(title) {
  return new Promise((resolve, reject) => {
    const url = `https://en.wikipedia.org/w/api.php?action=query&titles=${encodeURIComponent(title)}&prop=pageimages&format=json&pithumbsize=500`;
    https.get(url, { headers: { 'User-Agent': 'AntigravityBot/1.0' } }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          const pages = json.query.pages;
          const pageId = Object.keys(pages)[0];
          if (pages[pageId] && pages[pageId].thumbnail) {
            resolve(pages[pageId].thumbnail.source);
          } else {
            resolve(null);
          }
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

async function downloadImage(url, filename) {
  return new Promise((resolve, reject) => {
    const filePath = path.join(__dirname, 'public', 'images', filename);
    const file = fs.createWriteStream(filePath);
    https.get(url, { headers: { 'User-Agent': 'AntigravityBot/1.0' } }, (res) => {
      if (res.statusCode !== 200) {
        return reject(new Error(`Failed to download ${url}: ${res.statusCode}`));
      }
      res.pipe(file);
      file.on('finish', () => {
        file.close();
        resolve(filePath);
      });
    }).on('error', (err) => {
      fs.unlink(filePath, () => reject(err));
    });
  });
}

async function main() {
  for (const item of imagesToFetch) {
    try {
      console.log(`Fetching thumbnail for ${item.title}...`);
      const url = await fetchThumbnailUrl(item.title);
      if (url) {
        console.log(`Downloading ${url}...`);
        await downloadImage(url, item.filename);
        console.log(`Saved ${item.filename}`);
      } else {
        console.log(`No thumbnail found for ${item.title}`);
      }
    } catch (e) {
      console.error(`Error processing ${item.title}:`, e);
    }
  }
}

main();
